File: Smart-Bus/Smart-Bus/Classes/Main/ZoneControl/Mood/View/ShadeStatusButton/SHMoodShadeStatusButton.m
//
//  SHMoodShadeStatusButton.m
//  Smart-Bus
//
//  Created by Mark Liu on 2017/9/18.
//  Copyright © 2017年 SmartHome. All rights reserved.
//

#import "SHMoodShadeStatusButton.h"

@implementation SHMoodShadeStatusButton


- (instancetype)initWithFrame:(CGRect)frame {
    
    if (self = [super initWithFrame:frame]) {
        
        [self setUI];
    }
    return self;
}

- (void)awakeFromNib {
    
    [super awakeFromNib];
    
    [self setUI];
}

- (void)setUI {
    
    self.imageView.contentMode = UIViewContentModeScaleAspectFit;
}

- (void)layoutSubviews {
    
    [super layoutSubviews];
    
    CGSize titleSize = [self.titleLabel.text boundingRectWithSize:self.bounds.size options:NSStringDrawingUsesLineFragmentOrigin attributes:@{NSFontAttributeName: self.titleLabel.font} context:nil].size;
    
    self.titleLabel.frame_width = titleSize.width;
    self.titleLabel.preferredMaxLayoutWidth = self.titleLabel.frame_width;
    
    self.titleLabel.frame_x = (self.frame_width - self.titleLabel. frame_width) * 0.5;
    
    self.titleLabel.frame_y = 0;
    self.titleLabel.frame_height = self.frame_height;
    
    self.imageView.frame_x = self.titleLabel.frame_x + self.titleLabel.frame_width;
    
    self.imageView.frame_width = MIN(self.frame_width - self.imageView.frame_x, titleSize.height);
    
    self.imageView.frame_height = self.imageView.frame_width;
    
    self.imageView.frame_y = (self.frame_height - self.imageView.frame_height) * 0.5;
}

@end

File: Smart-Bus/Smart-Bus/Classes/Main/ZoneControl/Mood/View/ShadeStatusButton/SHMoodShadeStatusButton.swift
//
//  SHMoodShadeStatusButton.swift
//  Smart-Bus
//
//  Created by Apple on 2019/6/28.
//  Copyright © 2019 SmartHome. All rights reserved.
//

import UIKit

class SHMoodShadeStatusButton: UIButton {

    override init(frame: CGRect) {
        super.init(frame: frame)
        
        setUI()
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func awakeFromNib() {
        super .awakeFromNib()
        
        setUI()
    }

    
    private func setUI() {
        
        imageView?.contentMode = .scaleAspectFit
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        
        // 文字大小
        guard let titleSize =
            titleLabel?.text?.boundingRect(
                with: bounds.size,
                options: .usesLineFragmentOrigin,
                attributes: [
                    NSAttributedString.Key.font:
                        titleLabel?.font as Any
                ],
                context: nil
                ).size else {
                    
              return
        }
        
        // 文字
        titleLabel?.frame_y = 0
        titleLabel?.frame_height = titleSize.height
        
        titleLabel?.frame_width = titleSize.width
        titleLabel?.preferredMaxLayoutWidth = titleSize.width
        titleLabel?.frame_x =
            (frame_width - titleSize.width) * 0.5
        
        
        // 图片
        imageView?.frame_x =
            (titleLabel?.frame_x ?? 0) +
            (titleLabel?.frame_width ?? 0)
        
        imageView?.frame_width =
            CGFloat.minimum(
                frame_width - (imageView?.frame_x ?? 0),
                titleSize.height
        )
        
        imageView?.frame_height = imageView?.frame_width ?? 0
        
        imageView?.frame_y =
            (frame_height - (imageView?.frame_height ?? 0)) * 0.5
    }
}
